nearbridge/contracts/NearBridge.sol

```diff
@@ -44,25 +44,31 @@ contract NearBridge is INearBridge {
     address payable burner;
     uint256 public lockEthAmount;
     uint256 public lockDuration;
+    uint256 public replaceDuration;
     Ed25519 edwards;
 
     // Block producers of the current epoch.
     BlockProducerInfo public currentBlockProducers;
     // Block producers of the next epoch.
     BlockProducerInfo public nextBlockProducers;
-    // Backup current block producers. When candidate block is submitted and it comes from the next epoch, we backup
-    // current block producers. Then if it gets successfully challenged, we recover it the following way:
-    // nextBlockProducers <- currentBlockProducers
-    // currentBlockProducers <- backupCurrentBlockProducers
-    BlockProducerInfo public backupCurrentBlockProducers;
 
-    // The most recent block.
+    // The most recent head that is guaranteed to be valid.
     BlockInfo public head;
-    // The backup of the previous most recent block, in case it was challenged.
-    BlockInfo public backupHead;
 
-    mapping(uint64 => bytes32) public blockHashes;
-    mapping(uint64 => bytes32) public blockMerkleRoots;
+    // The most recently added block. May still be in its challenge period, so should not be trusted.
+    BlockInfo untrustedHead;
+    // True if untrustedHead is from the following epoch of currentHead.
+    // False if it is from the same epoch.
+    bool untrustedHeadIsFromNextEpoch;
+    // Next block producers from untrustedHead, if any.
+    BlockProducerInfo untrustedNextBlockProducers;
+    // Address of the account which submitted the last block.
+    address lastSubmitter;
+    // End of challenge period, or zero if there is no block to be challenged.
+    uint lastValidAt;
```

Comment:
Please add a comment saying something like: "when it is zero `untrustedHead`, `untrustedHeadIsFromNextEpoch`, `untrustedNextBlockProducers` should be considered to be not set".